packages/dashboard/src/components/negotiations-panel.tsx

```diff
@@ -275,17 +318,74 @@ export const NegotiationsPanel = React.memo((props: NegotiationsPanelProps) => {
     }
 
     updateNegotiationTrajectory();
+    setSelected(nodeIds);
+  };
+
+  const handleClearAllCurrNegotiations = () => {
+    localStorage.setItem('conflictIds', Object.keys(conflicts).toString());
+    setNegotiationContents({});
+  };
+
+  const handleResetNegotiations = () => {
+    setExpanded([]);
+    setSelected('');
+    setNegotiationTrajStore({});
+  };
+
+  const handleRestoreNegotiations = () => {
+    localStorage.setItem('conflictIds', '');
```

Comment:
A couple of comments on this:
I think this would great if the browser restarts suddenly but I don't know if we should be using localStorage here. In theory in the future, and please correct me if I'm the wrong @koonpeng , we should be consuming the history from the backend directly so we shouldn't be storing the conflicts. 

Some comments:
I saw that there are leftovers persisting after a refresh or a new session but it's not affecting the behavior though.
![image](https://user-images.githubusercontent.com/11761240/96492921-b3cdf100-121a-11eb-874b-054793e6e27c.png)

As an alternative, we can use a state (list or a stack) to store the conflicts in memory. But as you know this will disappear after a browser refresh.


For me, in the meanwhile, we can keep using the local storage to store that negotiation history, we can set the maximum quantity of negotiation to save too.

Reply:
I was also trying not to use local storage initially but `useState` is also extremely tricky to use for this case due to how it is designed.

> I saw that there are leftovers persisting after a refresh or a new session but it's not affecting the behavior though.

Yup, there are leftovers after a browser refresh but I guess since the existing conflicts are cleared, the stored values don't matter anymore.


> As an alternative, we can use a state (list or a stack) to store the conflicts in memory. But as you know this will disappear after a browser refresh.

Sounds like a good idea. Now that I think of it, since refreshing the browser will reset the existing conflicts, resetting this list also doesn't matter anymore

Reply:
I agree that local storage is not a good idea. I don't know how it gets the negotiations, why do we not want to clear the array? Even if so, can't we use state/memo/ref to persist the array across renders?

Reply:
> I agree that local storage is not a good idea. I don't know how it gets the negotiations, why do we not want to clear the array? Even if so, can't we use state/memo/ref to persist the array across renders?

Oops :facepalm: 

Ok so initially the confusion was with this piece of code here:

```
let negotiationContents: JSX.Element[] = [];
  if (conflicts) {
    let reversedConflicts = Object.keys(conflicts).reverse();
    reversedConflicts.forEach((version) => {
      const conflict = conflicts[version];
      let contents = renderNegotiations(version, conflict);
      negotiationContents.push(contents);
    });
  } else {
    console.log('prop negotationstatus empty');
  }
```

it wasn't in a function and was needed on the initial render so I can't use setState here. I tried putting it in `useEffect` but couldn't work. I then tried out all the above options, as you mentioned (useMemo, refs, useState) but all did not work too. I then refactored the negotiations to be an object in useState instead of an array hanging by itself.

Just tested saving the array in useState and it works now. Complete oversight from me and totally forgot the initial intentions of the refactor :sweat_smile:

Reply:
I think you should be able to use ref in this case

```tsx
const negotiationContentsRef = React.useRef<JSX.Element[]>([]);
const negotiationContents = negotiationContentsRef.current;
```

That should make it persist across renders.

Reply:
> I think you should be able to use ref in this case
> 
> ```tsx
> const negotiationContentsRef = React.useRef<JSX.Element[]>([]);
> const negotiationContents = negotiationContentsRef.current;
> ```
> 
> That should make it persist across renders.

Just tried it but I currently can't think of a way to re-render the component when I modify the value of ref, so I think I would still use useState for negotiationContents instead.